[#154938935] Implement Google Authorization with Firebase Authentication

asce1062 opened the pull request:
### What does this PR do?
Implement google authorization with firebase authentication

#### Description of Task to be completed?
- Ensure that a user can select a google account to use for logging
    in to the app.
- Ensure than the user is authenticated with firebase.
- Display the users details (Display Name, Email and Photo)
    on the navigation drawer.

#### What are the relevant pivotal tracker stories?
[#154938935](https://www.pivotaltracker.com/story/show/154938935)
[#155323924](https://www.pivotaltracker.com/story/show/155323924)
[#155325963](https://www.pivotaltracker.com/story/show/155325963)

#### Screenshots (if appropriate)
![art-mainactivity-03012018120252](https://user-images.githubusercontent.com/3494097/36835946-ed67c0e0-1d48-11e8-9a61-0dbf99e92d78.png)
![art-mainactivity-03012018120357](https://user-images.githubusercontent.com/3494097/36835947-ed95571c-1d48-11e8-8650-1dca6d4b28d6.png)
![art-mainactivity-03012018120432](https://user-images.githubusercontent.com/3494097/36835948-edc53c02-1d48-11e8-8749-8bbcbc186a65.png)

Commit:
feat(google-sign-in): implement google authorization with firebase authentication

- Ensure that a user can select a google account to use for logging
    in to the app.
- Ensure than the user is authenticated with firebase.
- Display the users details (Display Name, Email and Photo)
    on the navigation drawer.
- Set Roboto-Regular as the default font.
- Add the files for MVP.
- Change implementation statements to compile statements.
- Add Module annotation to relevant classes.
- Add MVP logic to view, presenter empty.
- Add butterknife and changes variable names.
- Use the Andela Theme Colors

[Delivers #154938935, #155325963 and #155323924]

zmwas commented on app/src/main/java/com/andela/art/Login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+        Window w = getWindow();
+        w.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
+        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
+        button = findViewById(R.id.googleSignInButton);
```
Use butterknife for binding views

mugiwaramunyi replied:
👍

zmwas commented on app/src/main/java/com/andela/art/Login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
+        button = findViewById(R.id.googleSignInButton);
+
+        button.setOnClickListener(new View.OnClickListener() {
```
Use butterknife for adding listeners

zmwas commented on app/src/main/java/com/andela/art/Login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+        mConnectionProgressDialog.setMessage("Signing in...");
+    }
+
+  @Override public void firebasewithGoogleAuth(GoogleSignInAccount account) {
```
This could be done in another layer and then injected into the activity so that the view remains dumb.

mugiwaramunyi replied:
Hey @zmwas in this instance this method has a listener on it. Is there a way of having listeners in a presenter?

zmwas replied:
You could create a class to do the login and then inject the class into the presenter. When login is called on the activity the presenter just uses the injected class and calls the sign in. That way the firebase code could be mocked as it will be injected in the constructor and the activity will remain dumb.

mugiwaramunyi replied:
Noted, that might work, will try this tomorrow.

zmwas commented on app/src/main/java/com/andela/art/Login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+  @Override public void firebasewithGoogleAuth(GoogleSignInAccount account) {
+    Log.d(TAG, "firebaseAuthWithGoogle:" + account.getId());
```
This Log should not be in the production code if you are done testing it.

muhallan commented on app/src/main/java/com/andela/art/MainActivity.java:
```diff
@@ -1,13 +1,23 @@
+ * Created by asce1062 on 23/02/2018.
+ */
 
 public class MainActivity extends AppCompatActivity {
```
This class seems to be doing nothing since LoginActivity is the first activity, it could be made the `launcher activity`

muhallan commented on app/src/main/java/com/andela/art/NavigationDrawer/NavigationDrawerActivity.java:
```diff
@@ -0,0 +1,135 @@
+
+    FirebaseAuth mAuth;
+    FirebaseAuth.AuthStateListener mAuthListener;
+    private TextView Name, Email;
```
Variables starting with capital letters is against the variable naming conventions; https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

muhallan reviewed (changes requested):
Github is complaining that some files do not have a blank space at the end. It should be so for every file pushed.

muhallan commented:
Please consider renaming `NavigationDrawerActivity` to `SerialEntryActivity` and the corresponding package to `serialentry` according to advice from @zmwas. 

Also the `Login` package should be in all lower case.

muhallan commented:
There seems to be an issue with the login. It says `Authorization failed: Ensure you have selected a Google account`.

muhallan commented on app/src/release/google-services.json:
```diff
@@ -0,0 +1,42 @@
+    }
+  ],
+  "configuration_version": "1"
+}
```
End of file line missing

asce1062 replied:
@muhallan This is auto generated by google after setting up an app. Do you still feel the need for an EOF line?

muhallan replied:
I think it can be added. Because as you can see Github puts a red marker on it

asce1062 replied:
EOF line added.

muhallan commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,7 @@
+    <dimen name="activity_vertical_margin">16dp</dimen>
+    <dimen name="nav_header_vertical_spacing">8dp</dimen>
+    <dimen name="nav_header_height">176dp</dimen>
+</resources>
```
End of file line missing

asce1062 replied:
FIxed

suada-haji commented on app/src/main/java/com/andela/art/login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        CalligraphyConfig.initDefault(new CalligraphyConfig.Builder()
```
You can directly set the fonts on your XML file, I don't think you need this library. Just create a font folder in the res directory and use font family to set it. https://developer.android.com/guide/topics/ui/look-and-feel/fonts-in-xml.html

asce1062 replied:
Hmm, this is only for API level >= 26. Our current minimum required API level is 22. Should we bump up the minimum required API level?

suada-haji replied:
No you don't need to bump up your API level. It's says in the documentation that `To use the Fonts in XML feature on devices running Android 4.1 (API level 16) and higher, use the Support Library 26.` Which you're already using

asce1062 replied:
Addressed.

suada-haji commented on app/src/main/java/com/andela/art/login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+        w.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
+        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
+
+        ButterKnife.bind(this);
```
With the latest android sdk you don't need Butterknife anymore, Android has its own Data Binding library, https://developer.android.com/topic/libraries/data-binding/index.html.
Let me know if you need assistance in implementing it

asce1062 replied:
Updated.

muhallan commented on app/src/main/res/layout/google_login.xml:
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="utf-8"?>
```
It would be better to name this as `activity_login` seeing it's the main layout for LoginActivity. It's just convention. 
https://hungsonandroid.wordpress.com/2016/10/17/androidnaming-convention/

asce1062 replied:
Addressed.

muhallan commented on app/src/main/java/com/andela/art/login/LoginActivity.java:
```diff
@@ -0,0 +1,175 @@
+                .setFontAttrId(R.attr.fontPath)
+                .build()
+        );
+        setContentView(R.layout.google_login);
```
This could point to `activity_login` xml. It would be more intuitive and that xml would be easy to find. Plus I think it's convention to do it like that. https://hungsonandroid.wordpress.com/2016/10/17/androidnaming-convention/

asce1062 replied:
Renamed.

Commit:
Set Roboto-Regular font for both the Navigation Header and Drawer Items

Commit:
Enable Databinding and remove Butterknife
    - Add the dataBinding element to build.gradle.
    - Update gradle.properties with Data Binding Compiler V2.

suada-haji commented on app/src/main/res/values/styles.xml:
```diff
@@ -25,4 +25,25 @@
     </style>
 
+    <style name="MenuItemsText" parent="@android:style/TextAppearance.Medium">
+        <item name="android:textColor">#000000</item>
```
You can use `@android:color/black` instead

suada-haji commented on app/src/main/res/values/styles.xml:
```diff
@@ -25,4 +25,25 @@
+    <style name="EmailText" parent="android:TextAppearance">
+        <item name="android:fontFamily">@font/roboto_regular</item>
+        <item name="android:textSize">@dimen/nav_drawer_text_size</item>
+        <item name="android:textColor">#B3FFFFFF</item>
```
This one can be added to the `colors` resource file and just reference it

Commit:
Make PR changes for fonts

suada-haji commented on app/src/main/res/values/styles.xml:
```diff
@@ -26,7 +26,7 @@
 
     <style name="MenuItemsText" parent="@android:style/TextAppearance.Medium">
-        <item name="android:textColor">#000000</item>
+        <item name="android:textColor">@android:color/black</item>
```
Great! Looks good

Commit:
Implement Data Binding

Commit:
Add instapk for sharing generated APK's

suada-haji commented on app/src/main/java/com/andela/art/login/LoginActivity.java:
```diff
@@ -0,0 +1,161 @@
+        Window w = getWindow();
+        w.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
+        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
+        ActivityLoginBinding activityLoginBinding = DataBindingUtil.setContentView(this, R.layout.activity_login);
```
Good implementation of Data binding 👍
